Adds first draft of Confirm a Deposit to Exchange

Annamarie2019 opened the pull request:
Adds the entire first draft so it can be seen and discussed in Files Changed.

Annamarie2019 committed:
Adds first draft of Confirm a Deposit to Exchange

Adds the entire first draft so it can be seen and discussed in Files Changed.

Annamarie2019 commented on confirm-deposit:
```diff
@@ -1 +1,109 @@
+The one-way hash prevents transactions from being reverted. 
+[Blocks](https://docs.flare.network/tech/glossary/) are added to the [blockchain](https://docs.flare.network/tech/glossary/) in chronological order and it may take a few blocks for the hash to be completed.
+To prevent problems, we confirm deposits several blocks back (five blocks in our code), when the blocks are most likely to be irreversible.
+
```
TL;DR: Do I characterize the -5 blocks correctly?

The [Architecture of an Exchange doc](https://docs.flare.network/exchange/architecture/) says, "To avoid problems, the Exchange should only act on transactions appearing on blocks old enough for the chance of them being reverted to be negligible."  Is this doc referring to the fact that blocks are irreversible except in the case of a severe attack when an ecosystem might agree to fork. Am I on the right track? 
I have never heard of a period where they could possibly be reverted, so I'm imagining that is the brief moment waiting for the cryptographic hash to be in place, yes?

segfaultxavi replied:
I explained this in the text, but feel free to ask again if it's not clear.

Annamarie2019 replied:
Got it.

Annamarie2019 commented on confirm-deposit:
```diff
@@ -1 +1,109 @@
+After five blocks have been created, valid deposits are labeled "confirmed."
+
+Run the javascript below to see pending and confirmed transactions and which block number they were added to.
+
```
TL;DR: I think we should have general education on the blockchain elsewhere, so I limited blockchain explanation to only what they need to know for this procedure. Here are some terms to add to the glossary.

| Term | Description |
| ----------- | ----------- |
| address | The address of a wallet where assets can be stored. Represented by a hash. |
| block header | Data about the block, including a timestamp of when the block was a created, a hash representing the address of the previous block, and a Merkle root hash (a cryptographic hash of all of the transactions included in the block) |
| block number | Blocks are numbered sequentially in chronological order of creation |
| hash | An output of a string of a fixed size derived from the transformation of data of any length. `tx.hash` is the hash of the transaction. It is generated by a one-way cryptographic algorithm, for which the original data cannot be retrieved by decryption. Hashing stores passwords and prevents fraudulent transactions and double spending in blockchain. |
| node | A programmed participant in the network that can interact with other nodes. Each node has different responsibilities, such as client or validator. |

segfaultxavi replied:
I totally agree. A general introduction to blockchain would be very nice to have, and it would define all these concepts.
At the same time, they could be added to the glossary.
But that's beyond the scope of this exercise, so don't worry.

Annamarie2019 commented on confirm-deposit:
```diff
@@ -1 +1,109 @@
+cd flare`
+
+See [the Flare repository on GitHub](https://github.com/flare-foundation/flare).
+
```
The Architecture page says "Your own instance with the test server" Is this relevant here or should we just stick to running in the terminal?

segfaultxavi replied:
Not sure what you're asking. Which page says that? I don't see that sentence here: https://docs.flare.network/exchange/architecture

Annamarie2019 replied:
Sorry, that looks like some corruption of "Use your own node URL." (Or maybe an unintended cut & paste.)

My question is about "Use your own node URL." So, I guess I can't try that without spinning up a node. No problem.

But...In IPFS/Filecoin, a node has a peer Identifier, which is not location-based, sometimes called a cryptographic ID. is "URL" the correct term here?

segfaultxavi replied:
Ah, now I see it, sorry, it's the comment in the code.
Yeah, it means that the URL provided in the code is for testing purposes and in production you should spin up your own node and use its URL. It is not an ID, it is literally a URL, used to access the node, so that's the correct term here.
As you say, you would need to spin up your own node to test it, so don't worry.

Annamarie2019 replied:
Oh. OK. And, of course, wss://coston-api.flare.network/ext/bc/C/ws is an URL.

Annamarie2019 commented on confirm-deposit:
```diff
@@ -1 +1,109 @@
+`./scripts/build.sh`
+
+## Run the Code
+
```
How do they run this file? I've seen people run files in the cli, but I don't know how to do it. I created a file and ran `npm run confirm-deposit.js`, but it didn't work.

segfaultxavi replied:
It should work with `node confirm-deposit.js`.
I am no expert on node or js, so I am not sure what `npm run...` does.
In the future, whenever something does not work, give a bit more of information (what you typed and what the output was, for example) :)

Annamarie2019 replied:
Sorry about the dearth of information. It was due to a temporary lapse in faith, bc I didn't have success earlier. 

I think I did not have web3 (required!) installed and I had the wrong syntax `node run confirm-deposit.js` 

At any rate `node confirm-deposit.js` worked fine today.

segfaultxavi replied:
:+1:

Annamarie2019 commented on confirm-deposit:
```diff
@@ -1 +1,109 @@
+* Sets up the deposit by instantiating `web3`, assigning the endpoint URL, and assigning the `receivingAddress`
+* Subscribes to 'pending transactions` to listen for deposit transactions and, if they are sent to the receiving address, labels them as "pending."
+* Subscribes to 'newBlockHeaders` to get a block that is five blocks back from the block with the pending transactiions and labels it as "confirmed."
+
```
Since multiple transactions may be added to a block at a time, are we marking 1 pending transaction at a time, and then the entire set (1 or more) of confirmed transactions in a block?

segfaultxavi replied:
This is exactly what the code does, yes!
The `pendingTransactions` event is received for each individual transaction, whereas the `newBlockHeaders` event retrieves a block, and then iterates over all the transactions contained in it (the `forEach` loop).

Annamarie2019 replied:
Of course! Now I see: `getTransaction` and `forEach`!

Annamarie2019 commented on confirm-deposit:
```diff
@@ -1 +1,109 @@
+// You must have web3 installed 
+const Web3 = require('web3');
+
+// Test this code on the Coston testnet. Use your own node URL for actual runtime.
```
"Use your own node URL": They can test it as is, yes? Their own node URL would be for actual run time, yes?

segfaultxavi replied:
Correct.

Annamarie2019 commented on confirm-deposit:
```diff
@@ -1 +1,109 @@
+
+* Sets up the deposit by instantiating `web3`, assigning the endpoint URL, and assigning the `receivingAddress`
+* Subscribes to 'pending transactions` to listen for deposit transactions and, if they are sent to the receiving address, labels them as "pending."
+* Subscribes to 'newBlockHeaders` to get a block that is five blocks back from the block with the pending transactiions and labels it as "confirmed."
```
Oops! Transactiions has a typo.

segfaultxavi replied:
Happens to the best of us :D
You better install a spell checker to your markdown editor...

Annamarie2019 commented on confirm-deposit:
````diff
@@ -1 +1,109 @@
+    });
+}).on("error", console.error);
+```
+
````
Do we have more information on possible error codes, such as [Transaction Status](https://github.com/flare-foundation/multi-chain-client/blob/main/docs/definitions/transaction-status.md)

segfaultxavi replied:
Good point. It is always a good idea to talk about return values and error codes. Otherwise, when things do not work users are lost and get very frustrated.
In this case, it would be nice to add that link, but we should check that's actually what's being returned. I am not sure right now.

Annamarie2019 replied:
This is from multi-chain client: `Success` doesn't seem applicable, bc we give the "pending" and "confirmed" statuses. However, `failed due to sender` and `failed due to receiver` sound plausible and may be worth looking into.

| ID  | Status                      | Description                                                                                                                                                        |
| --- | --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `0` | **Success**                 | Transaction was successful and payment was made.                                                                                                                   |
| `1` | **Failure due to sender**   | Transaction failed but it was still included in a block and the fee was paid. The failure cannot be clearly attributed to the receiver.                            |
| `2` | **Failure due to receiver** | Transaction failed but it was still included in a block and the fee was paid. The failure can be clearly attributed to the receiver (e.g. bad destination address) |

Annamarie2019 commented:
@segfaultxavi Here is the first draft with 7 questions.

Annamarie2019 committed:
Rename confirm-deposit to confirm-deposit.md

Annamarie2019 committed:
Delete deposit-to-exchange.md

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,109 @@
+One of the stages is to go through the [Exchange server](https://docs.flare.network/exchange/architecture/#architecture-of-an-exchange), which monitors for deposits and updates the Balances database. 
+Run the code below to have the Exchange server listen for deposit transactions and confirm the deposits.
+
+On a blockchain, multiple [validators](h[ttps://docs.flare.network/tech/glossary/](https://docs.flare.network/tech/validators/)) examine newly created transactions and come to an agreement about whether they are valid. 
```
```suggestion
On a blockchain, multiple [validators](https://docs.flare.network/tech/validators/) examine newly created transactions and come to an agreement about whether they are valid. 
```

Be careful!

Have you tried actually rendering the markdown to produce an HTML page? You would very easily see that this does not show up correctly.

Annamarie2019 replied:
I will try that today....and add a correct link.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,109 @@
+
+On a blockchain, multiple [validators](h[ttps://docs.flare.network/tech/glossary/](https://docs.flare.network/tech/validators/)) examine newly created transactions and come to an agreement about whether they are valid. 
+When they are valid, they are added to a new block on the chain and identified by a one-way cryptographic algorithm, called a _hash_. 
+The one-way hash prevents transactions from being reverted. 
```
Not quite... the hash is a _signature_, so what it does is prevent the block from being tampered with after it has been added to the blockchain.

Annamarie2019 replied:
Got it. I wrote _hash_ out of the intro and focused on validators.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,109 @@
+On a blockchain, multiple [validators](h[ttps://docs.flare.network/tech/glossary/](https://docs.flare.network/tech/validators/)) examine newly created transactions and come to an agreement about whether they are valid. 
+When they are valid, they are added to a new block on the chain and identified by a one-way cryptographic algorithm, called a _hash_. 
+The one-way hash prevents transactions from being reverted. 
+[Blocks](https://docs.flare.network/tech/glossary/) are added to the [blockchain](https://docs.flare.network/tech/glossary/) in chronological order and it may take a few blocks for the hash to be completed.
```
Again, not exactly. Calculating a hash is immediate, what takes time is for all the validators to agree on which should be the next block to be added.
Think of a room full of people trying to agree on anything... it takes some time.

Annamarie2019 replied:
Got it.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,109 @@
+When they are valid, they are added to a new block on the chain and identified by a one-way cryptographic algorithm, called a _hash_. 
+The one-way hash prevents transactions from being reverted. 
+[Blocks](https://docs.flare.network/tech/glossary/) are added to the [blockchain](https://docs.flare.network/tech/glossary/) in chronological order and it may take a few blocks for the hash to be completed.
+To prevent problems, we confirm deposits several blocks back (five blocks in our code), when the blocks are most likely to be irreversible.
```
And the final piece of the puzzle. Waiting for 100% agreement of all validators would be too slow (imagine you pay with your credit card on a shop and you are waiting for the payment to be confirmed. You want this to be quick!).
Therefore, transactions are usually accepted **before** there's full consensus. This means that people on different parts of the room might reach different conclusions on whether your transaction was accepted or not. These are called different _branches_ of the blockchain.
Later, since these people are constantly moving around the room and talking to each other, some might realize that different conclusions were reached for a past block and they'll fix it. Fixing it means that one of the branches is discarded and its followers jump to the other one. And in turn, this means that some transactions that had already been accepted in the discarded branch are now not accepted anymore. THIS is where reverted transactions come from.
And the way to avoid this is to wait for a few blocks, so that all validators have had a chance to look at your transaction and you are sure it will not be reverted.

Annamarie2019 replied:
Awesome, I'm learning some more detail about how validators work.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,109 @@
+[Blocks](https://docs.flare.network/tech/glossary/) are added to the [blockchain](https://docs.flare.network/tech/glossary/) in chronological order and it may take a few blocks for the hash to be completed.
+To prevent problems, we confirm deposits several blocks back (five blocks in our code), when the blocks are most likely to be irreversible.
+
+Newly submitted deposits to our receiving address are labeled "pending."
```
I always try to avoid talking about "we", "us" or "our".

Annamarie2019 replied:
I agree. Sometimes that takes us to passive voice, which is also evil.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,109 @@
+2. Clone the Flare repository: 
+
+`git clone https://github.com/flare-foundation/flare.git
+cd flare`
```
Code snippets should use code blocks like this:

```sh
git clone https://github.com/flare-foundation/flare.git
cd flare
```

Single backquotes should be used only for inline literal text, when referring to `variables` or `program_names`, for example.

Annamarie2019 replied:
Makes sense.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,109 @@
+The following code:
+
+* Sets up the deposit by instantiating `web3`, assigning the endpoint URL, and assigning the `receivingAddress`
+* Subscribes to 'pending transactions` to listen for deposit transactions and, if they are sent to the receiving address, labels them as "pending."
```
Be careful, the quotes used around `pending transactions` do not match so they break the format. Again, rendering the markdown into HTML would easily show you that there is a problem.

Annamarie2019 replied:
Got it.

segfaultxavi commented on confirm-deposit.md:
````diff
@@ -0,0 +1,109 @@
+
+For example,
+
+```javascript
````
Well, this code block is not javascript. But we're not strict about this, this is only used to produce syntax coloring so you are free to use whatever "looks good".

Annamarie2019 replied:
Obvioulsy not a language I know. What is it? Typescript?

segfaultxavi replied:
It's isn't any language, just console output. So the language tag should be just `text`, meaning that no syntax highlighting will be used.
That said, sometimes, if the console output _resembles_ some other language (like js), it's worth using `javascript` instead of `text` so you get the numbers colored differently, for example, which is nice.

Annamarie2019 replied:
Got it. So I:
* changed mention of the code to "code," instead of "javascript."
* Gave the file suggestion as <name-of-file>.txt (It runs fine.)
* But I think "javascript" for the code block and outcome helps the reader (especially the grayed-out comments).

segfaultxavi replied:
Oh, we didn't understand each other! The code starting in line 61 is definitely javascript!
What is not javascript is the console output starting in line 120, that's what my initial comment referred to.
So you can rename the file to `.js`. `.txt` will work, but it's misleading.

segfaultxavi replied:
👍

Annamarie2019 committed:
Edits intro, "The following code" section & more

- Adds validators-agreeing rationale to explanation of -5 in code (still needs that blocks are ultimately immutable)
- Starts idea for run the code (still needs file name for test)
- Edits Subscribes to `newBlockHeaders` bullet point
- Changes "our" to "the" receiving address in code comments
- Fixes spelling and code formatting

Annamarie2019 committed:
Adds line space after 1. Run the code:

...to put it on a separate line.

Annamarie2019 committed:
Adds tip to check web3 installation

`npm web3 -v`

Annamarie2019 committed:
Adds link for next step

Annamarie2019 committed:
Edits in response to preliminary review

- Adds context of where this process starts. (Balances end of article where I add context of where it goes next.)
- Wordsmiths the reason for the five blocks back.
- Moves explanation of code ("The following code:" above "To run the code,"
- Adds the "Run `node <name of file>.js`" step. (I ran it and works fine.)
- Grabbed one of the results from running the file for the Outcome.

Annamarie2019 committed:
Wordsmiths validator paragraph

Annamarie2019 commented on confirm-deposit.md:
```diff
@@ -0,0 +1,120 @@
+On a [blockchain](https://docs.flare.network/tech/glossary/), multiple [validators](https://docs.flare.network/tech/validators/) examine newly created transactions and decide on whether they are valid and which [block](https://docs.flare.network/tech/glossary/) to add to the chain next. 
+It takes time for all the validators to agree.
+Until they reach agreement, validators may add or revert transactions to a block. 
+After the validation process is complete, reversing a transaction is very unlikely. 
```
I see from the page on validators that probabilities of being a leader are involved in the consensus algorithm. Is there a point at which transactions on a block become irreversible/immutable?

segfaultxavi replied:
The description of the consensus you are giving here is generic, and I think that is OK. I don't think there's any need to go into the details of how the consensus works for this page. So no need to talk about leaders.

Regarding immutability of transactions, some blockchains implement "Finality", which is a process that guarantees that a transaction won't change. But not us (nor Ethereum, for instance), so there is no such guarantee.
ANY transaction can be reverted in the future, but the probability of that happening is reduced with each new block. Therefore after a certain number of blocks it is commonly accepted that the transaction is "secure". The exact number of blocks depends on the blockchain and the security you want to provide: it's a tradeoff between security and speed.
The example uses 5 blocks, I think Bitcoin and Ethereum use 6 blocks, and some Exchanges might use 12 blocks, just to be extra secure.

Annamarie2019 replied:
Thanks.

Just to make sure (for my own understanding....no change in text), it sounds like you're saying that validators can revert a transaction, but it is understood that blocks are still tamper-resistant from malicious actors (by virtue of cryptography).

Annamarie2019 committed:
Edits first paragraph

Annamarie2019 committed:
Adds spaces between sentences.

Annamarie2019 committed:
Adds spacing to second lines of steps

...In One-Time Setup

Annamarie2019 committed:
Moves steps 2 and 3 or Run code

They belong after the code.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,120 @@
+Run the code below to have the Exchange server listen for deposit transactions and confirm the deposits.
+
+On a [blockchain](https://docs.flare.network/tech/glossary/), multiple [validators](https://docs.flare.network/tech/validators/) examine newly created transactions and decide on whether they are valid and which [block](https://docs.flare.network/tech/glossary/) to add to the chain next. 
+It takes time for all the validators to agree. 
```
I know that's personal, but do you like short, disconnected sentences like this one? I prefer to connect them, e.g.: "Since it takes time for all validators to agree, there is a period of time in which they might add or revert transactions to a block."
This is up for discussion, obviously, but it needs to be agreed upon!

Annamarie2019 replied:
Much better. (The short sentences are me figuring out the logic and what to leave out to keep it simple.)

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,120 @@
+It takes time for all the validators to agree. 
+Until they reach agreement, validators may add or revert transactions to a block. 
+After the validation process is complete, reversing a transaction is very unlikely. 
+To avoid the risk of confirming a block that could be reverted, confirm deposits several blocks back (for example, five blocks back as in the code below). 
```
```suggestion
To avoid the risk of accepting a deposit from a block that could be later reverted, only transactions that are a few blocks old are considered (for example, five blocks back as in the code below). 
```

I think this is clearer? This code is not confirming any block, since this is not code to be run by a validator, this is code for an Exchange. We're just acting upon blocks here.

Annamarie2019 replied:
Thanks for the nuance.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,120 @@
+Older blocks are less likely to be reverted.
+
+Newly submitted deposits to the receiving address are labeled "pending."
+After five more blocks have been created, valid deposits are labeled "confirmed."
```
```suggestion
After five more blocks have been created, valid deposits are then labeled "confirmed."
```

Again, I like creating a story rather than having disconnected sentences. Up for debate.

Annamarie2019 replied:
I was thinking that "after" clarifies the timeline, but developers' first language may be have them thinking it the "if-then" pattern. So, agreed. 
(In general, I try to avoid redundancy, but sometimes it is good for emphasis and helps the reader. Notice I use the phrase "your own" on line 65...that's to help them hear it more clearly.)

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,120 @@
+
+3. Install the web3.js library: 
+
+   `npm install web3`
```
Remember to use single backquotes only in inline text. This is a standalone code block so it should use triple backquotes (as in the `git clone` block above).

Annamarie2019 replied:
That makes sense! (My reviewers in the past have let me get away with this!) Much better.... Any standalone code, not just more than 1 line.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,120 @@
+
+The following code:
+
+* Sets up the deposit by instantiating `web3`, assigning the endpoint URL, and assigning the `receivingAddress`
```
Missing period.

Annamarie2019 replied:
I went back and forth about that. Sometimes I leave a period out after code, so it doesn't get confused with the code. In this case it's really clear, so agreed.

Annamarie2019 replied:
I went back and forth about that. Sometimes I leave a period out after code, so it doesn't get confused with the code, but it's really clear here, so agreed.

segfaultxavi commented on confirm-deposit.md:
```diff
@@ -0,0 +1,120 @@
+The following code:
+
+* Sets up the deposit by instantiating `web3`, assigning the endpoint URL, and assigning the `receivingAddress`
+* Subscribes to `pending transactions` to listen for deposit transactions and, if they are sent to the receiving address, labels them as "pending."
```
It's actually `pendingTransactions`. The backquotes mean this is literal text, so it should match the code exactly.

Annamarie2019 replied:
Typo sorry.

segfaultxavi reviewed (changes requested):
Looking good!

Annamarie2019 committed:
Fixes typo and block code ticks

* Final closing parentheses
* Tick marks for block code for steps 3 & 4 for One-Time Setup and step 3 of Run the Code.

Annamarie2019 committed:
Fixes Outcome heading to H2

Annamarie2019 committed:
Changes "the Exchange" to "an exchange"

This is generic, not specific to flare.

Annamarie2019 committed:
Reverts file type to .js

lines 59 and 111

Annamarie2019 commented:
@segfaultxavi Thanks! Now I feel much better. That makes sense!

Annamarie2019 committed:
Update confirm-deposit.md

Annamarie2019 committed:
Update confirm-deposit.md

Annamarie2019 committed:
Update confirm-deposit.md

Annamarie2019 committed:
Update confirm-deposit.md

Annamarie2019 committed:
Update confirm-deposit.md
